qiskit/quantum_info/operators/op_shape.py

```diff
@@ -0,0 +1,496 @@
+                return False
+        elif self._dims_r or self._num_qargs_r:
+            if raise_exception:
+                raise QiskitError('Input dimension should be empty for vector shape.')
+            return False
+
+        return True
+
+    @classmethod
+    def auto(cls, shape=None, dims_l=None, dims_r=None, dims=None,
+             num_qubits_l=None, num_qubits_r=None, num_qubits=None):
+        """Construct TensorShape with automatic checking of qubit dimensions"""
+        if dims and (dims_l or dims_r):
+            raise QiskitError(
+                "`dims` kwarg cannot be used with `dims_l` or `dims_r`")
+        if num_qubits and (num_qubits_l or num_qubits_r):
+            raise QiskitError(
+                "`num_qubits` kwarg cannot be used with `num_qubits_l` or `num_qubits_r`")
+
+        if num_qubits:
+            num_qubits_l = num_qubits
+            num_qubits_r = num_qubits
+        if dims:
+            dims_l = dims
+            dims_r = dims
+
+        if num_qubits_r and num_qubits_l:
+            matrix_shape = cls(num_qargs_l=num_qubits_r,
+                               num_qargs_r=num_qubits_l)
+        else:
+            ndim = len(shape) if shape else 0
+            if dims_r is None and num_qubits_r is None and ndim > 1:
+                dims_r = shape[1]
+
+            if dims_l is None and num_qubits_l is None and ndim > 0:
+                dims_l = shape[0]
+
+            if num_qubits_r is None:
+                if isinstance(dims_r, Integral):
+                    nq = int(log2(dims_r))
+                    if dims_r == 2 ** nq:
+                        num_qubits_r = nq
+                        dims_r = None
+                    else:
+                        dims_r = (dims_r, )
+                elif dims_r is not None:
+                    if set(dims_r) == {2}:
+                        num_qubits_r = len(dims_r)
+                        dims_r = None
+                    else:
+                        dims_r = tuple(dims_r)
+
+            if num_qubits_l is None:
+                if isinstance(dims_l, Integral):
+                    nq = int(log2(dims_l))
+                    if dims_l == 2 ** nq:
+                        num_qubits_l = nq
+                        dims_l = None
+                    else:
+                        dims_l = (dims_l, )
+                elif dims_l is not None:
+                    if set(dims_l) == {2}:
+                        num_qubits_l = len(dims_l)
+                        dims_l = None
+                    else:
+                        dims_l = tuple(dims_l)
+
+            matrix_shape = cls(dims_l=dims_l, dims_r=dims_r,
+                               num_qargs_l=num_qubits_l,
+                               num_qargs_r=num_qubits_r)
+        # Validate shape
+        if shape:
+            matrix_shape.validate_shape(shape)
+        return matrix_shape
+
+    def subset(self, qargs=None, qargs_l=None, qargs_r=None):
```

Comment:
is this function needed? where is it used? is it needed in future applications?

Reply:
It's basically the complement of the `remove` function. It's not used right now, but I think it might be useful for future applications